### PR TITLE
Normalization

### PR DESCRIPTION
**Pull request by javdrher**
Following the discussion in #7 , this provides the normalizer object appearing in that issue. The normalizer wraps a GPflow model and transparently scales data going in, and data/predictions going out. This means the scalings are only visible to the models, the rest of GPflowOpt sees the domain as specified by the user, and the output as obtained by the expensive function.

- Input is scaled from the user domain to [0,1]
- Output is normalized.

As the acquisition objects are unaware of the domain, the normalizer is switched off by default at wrapping time. When the acquisition object ends up in a BayesianOptimizer, the domain is available and if scaling is requested (True by default), the normalizers are enabled.

Before this can be merged, documentation is needed. However, I suggest to do this after merging #22 .

### CODE REVIEW

File: GPflowOpt/domain.py [hunk 1]
```diff
@@ -188,3 +188,10 @@ def _html_table_rows(self):
         Html row representation of a ContinuousParameter.
         """
         return "<tr><td>{0}</td><td>{1}</td><td>{2}</td></tr>".format(self.label, 'Continuous', str(self._range))
+
+
+class UnitCube(Domain):
+    def __init__(self, n_inputs):
+        params = [ContinuousParameter('u{0}'.format(i), 0, 1) for i in np.arange(n_inputs)]
```
Comment (javdrher) [inline GPflowOpt/domain.py]: Not really :)

File: GPflowOpt/acquisition.py [hunk 2]
```diff
@@ -72,28 +75,27 @@ def _optimize_models(self):
...
-            return acq, tf.gradients(acq, [Xcand], name="acquisition_gradient")[0]
-        else:
-            return acq
-
     def build_acquisition(self):
         raise NotImplementedError
 
+    def enable_scaling(self, domain):
+        """
+        Enables and configures the :class:`.Normalizer` objects wrapping the GP models.
```
Comment (icouckuy) [inline GPflowOpt/acquisition.py]: DataScaler

File: GPflowOpt/acquisition.py [hunk 3]
```diff
@@ -72,28 +75,27 @@ def _optimize_models(self):
...
+        :param domain: :class:`.Domain` object, the input transform of the normalizers is configured as a transform from
```
Comment (icouckuy) [inline GPflowOpt/acquisition.py]: normalizers -> data scalers

File: GPflowOpt/bo.py [hunk 4]
```diff
@@ -28,24 +28,31 @@ class BayesianOptimizer(Optimizer):
...
         :param optimizer: (optional) Optimizer object used to optimize acquisition. If not specified, SciPyOptimizer
          is used. This optimizer will run on the same domain as the BayesianOptimizer object.
-        :param initial: (optional) Design object used as initial set of candidates evaluated before the optimization 
-         loop runs. Note that if the underlying data already contain some data from an initial design, this design is 
-         evaluated on top of that.
+        :param initial: (optional) Design object used as initial set of candidates evaluated before the optimization
+         loop runs. Note that if the underlying model already some data from an initial design, it is augmented with the
+         evaluations obtained by evaluating the points as specified by the design
+        :param scaling: (boolean, default true) if set to true, the outputs are normalized, and the inputs are
+          scaled to :class:`.Normalizer`. This only affects model training: calls to acquisition.data, as well as
```
Comment (icouckuy) [inline GPflowOpt/bo.py]: DataScaler

File: GPflowOpt/scaling.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+    Model-wrapping class, primarily intended to assure the data in GPflow models is scaled. One DataScaler wraps one
+    GPflow model, and can scale the input as well as the output data. By default, if any kind of object attribute
+    is not found in the datascaler object, it is searched on the wrapped model.
+
+    The datascaler supports both input as well as output scaling, although both scalings are set up differently:
+
+    - For input, the transform is not automatically generated. By default, the input transform is the identity
+      transform. The input transform can be set through the setter property, or by specifying a domain in the
+      constructor. For the latter, the input transform will be initialized as the transform from the specified domain to
+      a :class:`.UnitCube`. When X is updated, the transform does not change.
```
Comment (icouckuy) [inline GPflowOpt/scaling.py]: to a UnitCube
?

File: GPflowOpt/scaling.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+        If enabled, data will be automatically scaled to zero mean and unit variance. When the output normalization is
+        switched on or off, the data in the model is automatically adapted.
+        :param flag: boolean, turn output scaling on or off
+        """
+
+        self._normalize_Y = flag
+        if not flag:
+            self.output_transform = LinearTransform(np.ones(self.Y.value.shape[1]), np.zeros(self.Y.value.shape[1]))
+        else:
+            self.Y = self.Y.value
```
Comment (icouckuy) [inline GPflowOpt/scaling.py]: I assume this trigger setattr which looks at the previously set self._normalize_Y

shouldnt output_transform be cleared or set to normalization?

File: GPflowOpt/scaling.py [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+        self.wrapped.X = self.input_transform.forward(value)
+
+    @Y.setter
+    def Y(self, value):
+        """
+        Set the output data. In case normalize_Y=True, the appropriate output transform is updated. It is then
+        applied on the data before setting the data of the wrapped model.
+        """
+        if self.normalize_output:
+            self.output_transform.assign(~LinearTransform(value.std(axis=0), value.mean(axis=0)))
```
Comment (icouckuy) [inline GPflowOpt/scaling.py]: ok

Comment (icouckuy) [inline GPflowOpt/scaling.py]: the assign bypasses setattr ? :)

File: GPflowOpt/scaling.py [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+        """
+        build_predict builds the TensorFlow graph for prediction. Similar to the method in the wrapped model, however
+        the input points are transformed using the input transform. The returned mean and variance are transformed
+        backward using the output transform.
+        """
+        f, var = self.wrapped.build_predict(self.input_transform.build_forward(Xnew), full_cov=full_cov)
+        return self.output_transform.build_backward(f), self.output_transform.build_backward_variance(var)
+
+    @AutoFlow((float_type, [None, None]))
+    def predict_f(self, Xnew):
```
Comment (icouckuy) [inline GPflowOpt/scaling.py]: dont we need only to override build_predict? GPFlow Model's predict_f will call our derived one?

File: GPflowOpt/scaling.py [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+        """
+        Compute the mean and variance of the latent function(s) at the points Xnew.
+        """
+        return self.build_predict(Xnew)
+
+    @AutoFlow((float_type, [None, None]))
+    def predict_f_full_cov(self, Xnew):
```
Comment (icouckuy) [inline GPflowOpt/scaling.py]: same

File: GPflowOpt/transforms.py [hunk 10]
```diff
@@ -93,38 +80,74 @@ def __init__(self, A, b):
...
+        self.b = DataHolder(b)
+
+    def build_forward(self, X):
+        return tf.matmul(X, tf.transpose(self.A)) + self.b
+
+    @AutoFlow((float_type, [None, None]))
     def backward(self, Y):
         """
-        Overwrites the default backward approach, it avoids an explicit matrix inversion.
+        Overwrites the default backward approach, to avoids an explicit matrix inversion.
```
Comment (icouckuy) [inline GPflowOpt/transforms.py]: approach to avoid an

File: GPflowOpt/transforms.py [hunk 11]
```diff
@@ -93,38 +80,74 @@ def __init__(self, A, b):
...
         return tf.transpose(XT)
 
+    def build_backward_variance(self, Yvar):
+        """
+        Additional method for scaling variance backward (used in :class:`.Normalizer`). Can process both the diagonal
+        variances returned by predict_f, as well as full covariance matrices.
+        :param Yvar: N x N x P or N x P
+        :return: Yvar scaled, same rank and dimensionality as input
+        """
+        rank = tf.rank(Yvar)
```
Comment (icouckuy) [inline GPflowOpt/transforms.py]: does Yvar has a static TF shape? Would make things easier.

As a sidenote, for prediction that is harder to do but I wish GPFlow would do that for like training data where the shape is known upfront in DataHolder.

File: testing/test_acquisition.py [hunk 12]
```diff
@@ -53,9 +53,9 @@ def test_result_shape(self):
             with self.acquisition.tf_mode():
                 tens = self.acquisition.build_acquisition(x_tf)
                 self.assertTrue(isinstance(tens, tf.Tensor), msg="no Tensor was returned")
-                tf_shape = tens.get_shape().as_list()
-                self.assertEqual(tf_shape[0], 50, msg="Tensor of incorrect shape returned")
-                self.assertTrue(tf_shape[1] == 1 or tf_shape[1] is None)
+                # tf_shape = tens.get_shape().as_list()
```
Comment (icouckuy) [inline testing/test_acquisition.py]: Is this a  leftover? if it is not needed, remove.

File: GPflowOpt/scaling.py (hunk 6 above)
Comment (javdrher) [inline GPflowOpt/scaling.py]: That happens there. If `flag` is true, the normalization is computed and set in the assignment of the property. If not, the output_transform is reset to the identity transform.

File: GPflowOpt/scaling.py (hunk 7 above)
Comment (javdrher) [inline GPflowOpt/scaling.py]: it mostly avoids a change, requiring recompilation of the graph

File: GPflowOpt/scaling.py (hunk 8 above)
Comment (javdrher) [inline GPflowOpt/scaling.py]: The thing is that the DataScaler is a parameterized, not a model

**Pull request merged**